Introduce new charging properties and BQ24190 support

rriveramcrus opened the pull request:
Adds several new properties to the charging API inherited from the Linux Power Supply Class
- CHARGE_TYPE
- HEALTH
- CONSTANT_CHARGE_CURRENT
- PRECHARGE_CURRENT
- CHARGE_TERM_CURRENT
- CONSTANT_CHARGE_VOLTAGE

Adds support for the BQ24190 charging IC

rriveramcrus committed:
drivers: charger: Introduces BQ24190 support

Adds support for the BQ24190 family of charging ICs.

Signed-off-by: Ricardo Rivera-Matos <rriveram@opensource.cirrus.com>

rriveramcrus committed:
charger: Adds CHARGER_CHARGE_TYPE property

Adds a new runtime property called CHARGER_PROP_CHARGE_TYPE to
the charger API. This can be used to get/set the charging scheme
used to charge the battery pack.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
charger: Adds CHARGER_HEALTH property

Adds a new runtime property called CHARGER_PROP_HEALTH to
the charger API. This can be used to get the charger health
condition.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
charger: Adds CHARGER_CONSTANT_CHARGE_CURRENT property

Adds a new runtime property called CHARGER_PROP_CONSTANT_CHARGE_CURRENT
to the charger API. This can be used to get/set the charger constant
current sink target.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
charger: Adds CHARGER_PRECHARGE_CURRENT property

Adds a new runtime property called CHARGER_PROP_PRECHARGE_CURRENT
to the charger API. This can be used to get/set the charger
conditioning current sink target.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
charger: Adds CHARGER_CHARGE_TERM_CURRENT property

Adds a new runtime property called CHARGER_PROP_CHARGE_TERM_CURRENT
to the charger API. This can be used to get/set the charger
termination current target.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
charger: Adds CHARGER_CONSTANT_CHARGE_VOLTAGE property

Adds a new runtime property called CHARGER_PROP_CONSTANT_CHARGE_VOLTAGE
to the charger API. This can be used to get/set the charger
constant voltage regulation target.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
dts: battery: Create bindings for common battery properties

Adds a devicetree for describing common battery characteristics used
by multiple devices and subsystems.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
dts: charger: bq24190: Adds dt-bindings for BQ24190

Adds devicetree bindings for the BQ24190 family of charging ICs

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

marcusfolkesson commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,480 @@
+		return ret;
+	}
+
+	v = (v & BQ24190_REG_POC_CHG_CONFIG_MASK) >> BQ24190_REG_POC_CHG_CONFIG_SHIFT;
```
No need to shift down for this usage. But if you really want, have a look at the FIELD_GET macro.

rriveramcrus replied:
cool, thanks for the tip. ack.

marcusfolkesson reviewed (commented):
Generally looks good, I would prefer to replace many bit operations with FIELD_GET/FIELD_SET though.

teburd commented on drivers/charger/bq24190.h:
```diff
@@ -0,0 +1,171 @@
+/*
```
Is this likely to be used any other place than the bq24190.c file? If not then its perfectly fine to move this to the .c file.

rriveramcrus replied:
I could see `drivers/regulator` needing some of these defines for controlling the OTG and SYS rails. I think I'll leave the header as-is if this makes sense.

fabiobaltieri replied:
mmh not really, you certainly don't want to include a charger header from a regulator driver. Look at how other MFD drivers are written.

Either way, you may make an argument for the register definitions but the data/config ones should definitely go in the driver file if they don't have to be shared, just to keep the scope under control and for convenience.

rriveramcrus replied:
gotcha, well I think I'll move the data/config structs into `bq24190.c` and keep the defines in `bq24190.h`.

aaronemassey commented on include/zephyr/drivers/charger.h:
```diff
@@ -84,6 +98,66 @@ enum charger_status {
+ */
+enum charger_charge_type {
+	/** Charge type is unknown */
+	CHARGER_CHARGE_TYPE_UNKNOWN = 0,
```
As we discussed offline, let's try to make these and the other new enum values more explicitly defined. Probably starting with fewer of them too and we can add as needed.

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,485 @@
+static int bq24190_register_reset(const struct device *dev)
+{
+	const struct bq24190_config *const config = dev->config;
+	int ret, limit = 100;
```
that 100 could probably use a define and then a comment referring to whatever number led to this one in the datasheet

rriveramcrus replied:
This was inherited from the Linux driver. I couldn't find anything in the datasheet to support the reset logic's timing characteristics, so this 100 * 100µs, so this seems to be more of a colloquial value.

My previous tenure was supporting this part, and this polling routine and interval seems appropriate.

I will give the 100 tries a define and throw in a comment.

fabiobaltieri replied:
Oof, ok, this seems to be a pattern for this BQ parts, load of these in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/bq274xx/bq274xx.c as well, same story, some timings just copied from the Linux driver, some @nixward going full archeology on the datasheets. Feel free to take inspiration from that one, had quite few eyes on it.

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,485 @@
+
+	if (v & BQ24190_REG_F_NTC_FAULT_MASK) {
+		switch (v >> BQ24190_REG_F_NTC_FAULT_SHIFT & 0x7) {
+		case 0x1: /* TS1 Cold */
+		case 0x3: /* TS2  Cold */
+		case 0x5: /* Both Cold */
```
these should be define rather than comments

rriveramcrus replied:
ack

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,485 @@
+		val->health = CHARGER_HEALTH_OVERVOLTAGE;
+	} else if (v & BQ24190_REG_F_CHRG_FAULT_MASK) {
+		switch (v >> BQ24190_REG_F_CHRG_FAULT_SHIFT & 0x3) {
+		case 0x1: /* Input Fault (VBUS OVP or VBAT<VBUS<3.8V) */
```
same

rriveramcrus replied:
ack

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,485 @@
+		ss_reg = FIELD_GET(BQ24190_REG_SS_CHRG_STAT_MASK, ss_reg);
+
+		switch (ss_reg) {
+		case 0x0: /* Not Charging */
```
same

rriveramcrus replied:
ack

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,485 @@
+			    union charger_propval *val)
+{
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		return bq24190_charger_get_online(dev, val);
```
I don't think it make much sense to pass `val` in the union form to these functions, this should rather be a

```c
case CHARGER_PROP_ONLINE:
	return bq24190_charger_get_online(dev, &val->online);
```

Then in code review you can imediately see that the union has been resolved to the right type form the fact that `CHARGER_PROP_ONLINE` and `online` are next to each other and there's no way you can pick the wrong one in the function implementation.

rriveramcrus replied:
ack

fabiobaltieri commented on include/zephyr/drivers/charger.h:
```diff
@@ -169,6 +171,8 @@ union charger_propval {
 	/** CHARGER_PROP_HEALTH */
 	enum charger_health health;
+	/** CHARGER_PROP_CONSTANT_CHARGE_CURRENT */
+	uint32_t const_charge_curr;
```
I think we are leaving an opportunity on the table here: one of the problem I bumped into all the time with the sensor subsystem is that driver are messing up the scaling in the values. Here since we have the enums we can put the unit in the name.

How about `const_charge_curr_ua` and so on...? That would put the unit where it's needed in both the driver and application code rather than in the generic header and documentation that no reviewer remembers to check anyway.

@teburd @MaureenHelm @yperess @aaronemassey @gmarull @aasinclair

opinion?

rriveramcrus replied:
I am okay with this. ack.

aaronemassey replied:
I think that's a good idea! I'd almost always rather be explicit when possible.

rriveramcrus replied:
done

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -43,6 +43,10 @@ enum charger_property {
 	CHARGER_PROP_HEALTH,
+	/** Configuration of current sink used for charging in µA */
+	CHARGER_PROP_CONSTANT_CHARGE_CURRENT,
+	/** Configuration of current sink used for conditioning in µA */
```
spurious change

fabiobaltieri commented on include/zephyr/drivers/charger.h:
```diff
@@ -99,6 +183,18 @@ union charger_propval {
+	enum charger_charge_type charge_type;
+	/** CHARGER_PROP_HEALTH */
+	enum charger_health health;
+	/** CHARGER_PROP_CONSTANT_CHARGE_CURRENT */
+	uint32_t const_charge_curr_ua;
+	/** CHARGER_PROP_PRECHARGE_CURRENT */
+	uint32_t precharge_curr_ua;
+	/** CHARGER_PROP_CHARGE_TERM_CURRENT */
+	uint32_t charge_term_curr_ua;
+	/** CHARGER_PROP_CONSTANT_CHARGE_VOLTAGE */
+	uint32_t const_charge_volt_uv;
```
Ok but now the fields and property names are incoherent, for example the prop has "current" but the field is "curr_ua". I think that the two should match beside the prefix and case, what do you think?

So you'd have something like

```c

	case CHARGER_PROP_PRECHARGE_CURRENT_UA:
		return bq24190_charger_get_pc(dev, &val->precharge_current_ua);
```

and there would be no ambiguity between matching the property and the union field.

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,491 @@
+	case CHARGER_PROP_STATUS:
+		return bq24190_charger_get_status(dev, &val->status);
+	case CHARGER_PROP_CONSTANT_CHARGE_CURRENT:
+		return bq24190_charger_get_ccc(dev, &val->const_charge_curr_ua);
+	case CHARGER_PROP_CONSTANT_CHARGE_VOLTAGE:
+		return bq24190_get_cvc(dev, &val->const_charge_volt_uv);
+	case CHARGER_PROP_PRECHARGE_CURRENT:
+		return bq24190_charger_get_pc(dev, &val->precharge_curr_ua);
+	case CHARGER_PROP_CHARGE_TERM_CURRENT:
+		return bq24190_charger_get_tc(dev, &val->charge_term_curr_ua);
```
the first 4 have a full name (`get_charge_type`...) and these are all shortened (`get_pc`, `get_tc`...). Use the expanded name for everything for coherency?

rriveramcrus replied:
ack

fabiobaltieri commented on drivers/charger/bq24190.c:
```diff
@@ -0,0 +1,491 @@
+	return 0;
+}
+
+static int bq24190_get_cvc(const struct device *dev, uint32_t *voltage)
```
Keep the unit? `uint32_t voltage_uv` or something, same for the others. That way it's easier to review that the units are right.

rriveramcrus replied:
ack

fabiobaltieri commented on samples/charger/bq24190/src/main.c:
Does this need to be device specific? The whole point of the subsystem is to abstract the device, how about keeping this generic and making it print the properties that you'd expect most chargers to have?

rriveramcrus replied:
No, it's not device specific. 

I guess I don't really understand the scope of a sample app. I was under the assumption that most drivers should have one to demonstrate how a consumer would actually use it (sort of like the examples in devicetree bindings). If it's not required, I can just drop the sample app entirely.

fabiobaltieri replied:
Ideally there should be just one charging sample that demonstrates the charging API and could be used with any charger, so I think this should just be `sample/charger` or something like that, and we should avoid having a sample for every driver. What I'm saying is that the whole point of the API is to be generic for every charger, so there should be only one sample that covers the common characteristics for all of them if possible.

rriveramcrus replied:
I understand now, thanks for clarifying. I'll generify the sample app.

rriveramcrus replied:
Is 60a139c along the lines of what you had in mind?

fabiobaltieri replied:
yep, thanks

fabiobaltieri commented on samples/charger/bq24190/boards/nucleo_f401re.overlay:
```diff
@@ -0,0 +1,21 @@
+    status = "okay";
+    clock-frequency = <I2C_BITRATE_STANDARD>;
+
+    bq24190: bq24190@6b {
+        compatible = "ti,bq24190";
+        reg = <0x6b>;
+        status = "okay";
+
+        constant-charge-current-max-microamp = <1000000>;
+        constant-charge-voltage-max-microvolt = <4208000>;
+    };
```
If this is a shield then it should be in a shield definition, if it's just a custom setup you have for testing it should not be upstreamed. Instead there should be `build_all` test to test build all the drivers and that's it.

rriveramcrus replied:
do you have an example?

rriveramcrus replied:
Oh sorry, I meant an example of the tests, but I found plenty after reading your https://github.com/zephyrproject-rtos/zephyr/pull/62551#discussion_r1345989207. 

This is just a custom fly-wired setup from the eval board to a Nucelo F401RE eval board, so it sounds like it shouldn't be here.

fabiobaltieri replied:
Oh sorry the build_all bit, yeah load of them in `tests/drivers`, arguably every driver category should be there.

rriveramcrus committed:
samples: charger: Creates a generic sample application for chargers

Creates a generic sample application that executes a simple charging
task loop capable of completing a basic charge cycle.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

rriveramcrus committed:
tests: build_all: charger: Builds BQ24190 driver

Adds the BQ24190 driver to the build_all test.

Signed-off-by: Ricardo Rivera-Matos <ricardo.rivera-matos@cirrus.com>

fabiobaltieri commented on samples/charger/src/main.c:
```diff
@@ -0,0 +1,101 @@
+	int ret;
+
+	if (chgdev == NULL) {
+		printk("\nError: no charger device found.\n");
```
nit: why the leading `\n` here and below? printk normally only have a trailing one, there should not be anything on the console when this happens, if there is it should a problem of the other log/print

rriveramcrus replied:
ack, dropped the leading `\n`. It was not for any particular reason.

fabiobaltieri reviewed (commented):
just one more nitpick on the sample for me, thanks

aaronemassey commented on include/zephyr/drivers/charger.h:
```diff
@@ -37,6 +37,20 @@ enum charger_property {
+	/** Value should be of type enum charger_health */
+	CHARGER_PROP_HEALTH,
+	/** Configuration of current sink used for charging in µA */
+	CHARGER_PROP_CONSTANT_CHARGE_CURRENT_UA,
```
This is great! Definitely migrating the fuel gauge properties to this format.

teburd reviewed (approved):
looks good
